src/log-store/src/raft_engine/backend.rs

```diff
@@ -0,0 +1,548 @@
+            prev = engine_get(&engine, &key)?;
+        }
+        engine_put(&engine, key, value)?;
+        Ok(PutResponse { prev_kv: prev })
+    }
+
+    async fn batch_put(&self, req: BatchPutRequest) -> Result<BatchPutResponse, Self::Error> {
+        let BatchPutRequest { kvs, prev_kv } = req;
+        let mut batch = LogBatch::with_capacity(kvs.len());
+
+        let mut prev_kvs = if prev_kv {
+            Vec::with_capacity(kvs.len())
+        } else {
+            vec![]
+        };
+
+        for kv in kvs {
+            if prev_kv && let Some(kv) = engine_get(&self.engine.read().unwrap(), &kv.key)? {
+                prev_kvs.push(kv);
+            }
+            batch
+                .put(SYSTEM_NAMESPACE, kv.key, kv.value)
+                .context(RaftEngineSnafu)?;
+        }
+
+        self.engine
+            .read()
+            .unwrap()
+            .write(&mut batch, false)
+            .context(RaftEngineSnafu)?;
+
+        Ok(BatchPutResponse { prev_kvs })
+    }
+
+    async fn batch_get(&self, req: BatchGetRequest) -> Result<BatchGetResponse, Self::Error> {
+        let mut response = BatchGetResponse {
+            kvs: Vec::with_capacity(req.keys.len()),
+        };
+        let engine = self.engine.read().unwrap();
+        for key in req.keys {
+            let value = engine.get(SYSTEM_NAMESPACE, &key).unwrap();
+            response.kvs.push(KeyValue { key, value });
+        }
+        Ok(response)
+    }
+
+    async fn compare_and_put(
+        &self,
+        req: CompareAndPutRequest,
+    ) -> Result<CompareAndPutResponse, Self::Error> {
+        let CompareAndPutRequest { key, expect, value } = req;
+
+        let mut batch = LogBatch::with_capacity(1);
+        let engine = self.engine.write().unwrap();
+        let existing = engine_get(&engine, &key)?;
+        let eq = existing
+            .as_ref()
+            .map(|kv| kv.value == expect)
+            .unwrap_or(false);
+        if eq {
+            batch
+                .put(SYSTEM_NAMESPACE, key, value)
+                .context(RaftEngineSnafu)?;
+            engine.write(&mut batch, false).context(RaftEngineSnafu)?;
+        }
+        Ok(CompareAndPutResponse {
+            success: eq,
+            prev_kv: existing,
+        })
+    }
+
+    async fn delete_range(
+        &self,
+        req: DeleteRangeRequest,
+    ) -> Result<DeleteRangeResponse, Self::Error> {
+        let DeleteRangeRequest {
+            key,
+            range_end,
+            prev_kv,
+        } = req;
+
+        let range = RangeRequest {
+            key,
+            range_end,
+            limit: 0,
+            keys_only: true,
+        };
+        let range_resp = self.range(range).await?;
+
+        let mut prev_kvs = vec![];
+        let mut deleted = 0;
+        for kv in range_resp.kvs {
+            self.delete(&kv.key, false).await?;
+            if prev_kv {
+                prev_kvs.push(kv);
+            }
+            deleted += 1;
```

Comment:
If the key was absent at the first place, `deleted` should not plus one.

Reply:
> If the key was absent at the first place, `deleted` should not plus one.

We've already used `raneg` to list all keys within range, thus the key must be present.

Reply:
But the keys might be deleted just after you `range` them. For example, other delete operations ran concurrently.

Reply:
> But the keys might be deleted just after you `range` them. For example, other delete operations ran concurrently.

We can also consider this key as "deleted".

Reply:
We can, but it's not deleted by this delete operation, it's weird to record one to `deleted` which is returned to the caller.

Reply:
> We can, but it's not deleted by this delete operation, it's weird to record one to `deleted` which is returned to the caller.

There's no way to avoid concurrent delete unless we acquire the write lock here, which would be too expensive here when deleted field is no something that really matters consistency/visibility.